### PR TITLE
Make inline ABD predicates compatible with C++

### PR DESCRIPTION
**Pull request by freqlabs**
### Motivation and Context

FreeBSD's zfsd fails to build after e2af2acce3 due to strict type
checking errors from the implicit conversion between bool and boolean_t
in the inline predicate definitions in abd.h.

### Description

Use conditionals to return the correct value type from these functions.

### How Has This Been Tested?

Compiled zfsd on FreeBSD.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by ikozhukhov — APPROVED** [review]

**Comment by freqlabs** [comment]
For posterity, the error:
```
c++  -O2 -pipe -fno-common -DIN_BASE -I/usr/src/sys/contrib/openzfs/include -I/usr/src/sys/contrib/openzfs/lib/libspl/include -I/usr/src/sys/contrib/openzfs/lib/libspl/include/os/freebsd -I/usr/src/sys -include /usr/src/sys/contrib/openzfs/include/os/freebsd/spl/sys/ccompile.h -I/usr/src/cddl/usr.sbin -D_MACHINE_FLOAT_H_ -DHAVE_ISSETUGID -DNEED_SOLARIS_BOOLEAN -g -MD -MF.depend.zfsd_event.o -MTzfsd_event.o -Wno-format-zero-length -fstack-protector-strong -Wsystem-headers -Werror -Wall -Wno-format-y2k -Wno-uninitialized -Wno-unknown-pragmas -Wno-empty-body -Wno-string-plus-int -Wno-unused-const-variable -Wno-tautological-compare -Wno-unused-value -Wno-parentheses-equality -Wno-unused-function -Wno-enum-conversion -Wno-unused-local-typedef -Wno-address-of-packed-member -Wno-switch -Wno-switch-enum -Wno-knr-promoted-parameter -Qunused-arguments     -Wno-c++11-extensions   -c /usr/src/cddl/usr.sbin/zfsd/zfsd_event.cc -o zfsd_event.o
In file included from /usr/src/cddl/usr.sbin/zfsd/zfsd_event.cc:40:
In file included from /usr/src/sys/contrib/openzfs/include/sys/vdev_impl.h:31:
In file included from /usr/src/sys/contrib/openzfs/include/sys/bpobj.h:29:
In file included from /usr/src/sys/contrib/openzfs/include/sys/dmu.h:49:
In file included from /usr/src/sys/contrib/openzfs/include/sys/zio_compress.h:33:
/usr/src/sys/contrib/openzfs/include/sys/abd.h:180:9: error: cannot initialize return object of type 'boolean_t' with an rvalue of type 'bool'
        return ((abd->abd_flags & ABD_FLAG_LINEAR) != 0);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/sys/contrib/openzfs/include/sys/abd.h:186:9: error: cannot initialize return object of type 'boolean_t' with an rvalue of type 'bool'
        return ((abd->abd_flags & ABD_FLAG_LINEAR_PAGE) != 0);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/sys/contrib/openzfs/include/sys/abd.h:192:9: error: cannot initialize return object of type 'boolean_t' with an rvalue of type 'bool'
        return ((abd->abd_flags & ABD_FLAG_GANG) != 0);
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/src/cddl/usr.sbin/zfsd/zfsd_event.cc:54:
In file included from /usr/include/c++/v1/sstream:173:
In file included from /usr/include/c++/v1/ostream:142:
/usr/include/c++/v1/bitset:384:63: error: shift count >= width of type [-Werror,-Wshift-count-overflow]
        __r |= static_cast<unsigned long long>(__first_[__i]) << (sizeof(__storage_type) * CHAR_BIT);
                                                              ^  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
4 errors generated.
*** Error code 1

Stop.
make: stopped in /usr/src/cddl/usr.sbin/zfsd
```

**Review by ahrens — APPROVED** [review]

**Comment by ahrens** [comment]
I wasn't aware of this type-checking requirement, thanks for fixing it.  I wonder why this wasn't caught by buildbot.  Maybe it was masked by other failures on freebsd HEAD?  Or is this check not performed when doing the buildbot runs?

File: include/sys/abd.h [hunk 1]
```diff
@@ -177,19 +177,20 @@ abd_zero(abd_t *abd, size_t size)
 static inline boolean_t
 abd_is_linear(abd_t *abd)
 {
-	return ((abd->abd_flags & ABD_FLAG_LINEAR) != 0);
+	return ((abd->abd_flags & ABD_FLAG_LINEAR) != 0 ? B_TRUE : B_FALSE);
```
Comment (behlendorf) [inline include/sys/abd.h:180]:
```suggestion
	return ((abd->abd_flags & ABD_FLAG_LINEAR) ? B_TRUE : B_FALSE);
```

That's kind of unfortunate.  If we must return `B_TRUE` / `B_FALSE` I'd suggest dropping the `!= 0` bit which is redundant and I don't think helps readability in this case.

**Comment by freqlabs** [comment]
> I wasn't aware of this type-checking requirement, thanks for fixing it. I wonder why this wasn't caught by buildbot. Maybe it was masked by other failures on freebsd HEAD? Or is this check not performed when doing the buildbot runs?

It's not built with the CI because it's not in the ZFS tree. I encountered this while updating the vendor import in FreeBSD.

**Review by behlendorf — APPROVED** [review]

File: include/sys/abd.h (hunk 1 above)
Comment (freqlabs) [reply include/sys/abd.h]: Done 👍

**Pull request merged**